### ISSUE TITLE
Make type provider optional in CustomAttribute.DecodeValue

### ISSUE THREAD

**Issue by tmat** [dotnet/corefx#31889]
```C#
public struct CustomAttribute 
{
   public CustomAttributeValue<object> DecodeValue()
}
```

Common use case for attribute value decoding is to retrieve a literal, e.g. a string. To do so the `ICustomAttributeTypeProvider<TType> provider` parameter is not necessary. Currently the user of the API has to define a dummy provider, which clutters the code. We can define and use the dummy provider internally and return CustomAttributeValue without type information.

**Comment by tmat** [comment]
@nguerrera

**Comment by nguerrera** [comment]
I like this idea.

**Comment by nguerrera** [comment]
Actually, the type provider is also used to specify the type of arguments, not just the value of arguments of type `Type`. So it can't just throw or you won't even be able to get string literals.

Another issue is that you can't read anything with an enum value in it unless the provider resolves the enum underlying type to determine how many bytes to read.

**Comment by jnm2** [comment]
> Another issue is that you can't read anything with an enum value in it unless the provider resolves the enum underlying type to determine how many bytes to read.

This is something I have to solve over and over in every project that uses SR.Metadata. Would you consider adding an API which parses serialized type names into a type provider, and an API which resolves enum metadata in external assemblies given an assembly file resolver which the user would implement?

**Comment by tmat** [comment]
> an API which parses serialized type names into

Yes, decoding and encoding serialized type names would be useful. Could you file an issue?

> an API which resolves enum metadata in external assemblies given a user-implemented assembly file resolver

No, such API does not belong to System.Reflection.Metadata. 
https://github.com/dotnet/corefx/issues/2800 might be what you are looking for.

**Comment by jnm2** [comment]
@tmat 

> Yes, decoding and encoding serialized type names would be useful. Could you file an issue?

 opened: https://github.com/dotnet/corefx/issues/32067

> No, such API does not belong to System.Reflection.Metadata. #2800 might be what you are looking for.

Okay, probably. #2800 seems like a heavyweight way to deal with a narrow problem, but it could be more widely useful too.

**Comment by jnm2** [comment]
@nguerrera 

> Another issue is that you can't read anything with an enum value in it unless the provider resolves the enum underlying type to determine how many bytes to read.

It just sunk in that changing the underlying type of an enum is one of the worst possible breaking changes for a library author. Attribute decoders will have no option but to read too much or too little and mess up everything that follows!

Changing the underlying type of an enum doesn't get a mention at https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md. Should it?

**Comment by nguerrera** [comment]
It should get a mention, yes. Attribute underlying types must not be changed.
